Microsatellite mutations models

andrewkern opened the pull request:
Here is a PR adding a three specific microsatellite mutation models along with a more general backend to add more. 

The generalized parameterization I've chosen here follow from [this nice paper](https://academic.oup.com/genetics/article/168/1/383/6059470) from Rasmus Nielsen's group. It offers plenty of flexibility and room for growth. I've currently implemented just 3 of the models in that paper, but it will be trivial to add more. 

I've implemented these as a subclass of `MatrixMutationModel`, and so I'm translating the CTMC representation from the original paper into the scaled, jump process transition matrix that is used in `MatrixMutationModel`. I hope I got the details correct here but may be off by a scalar?

Hoping to get some feedback as to what others would like to see here and what clean up to do. 

i should add-- this would close #2013

jeromekelleher commented:
@petrelharp - any thoughts on the above? What do you think of my ideas for the interface?

andrewkern commented on msprime/__init__.py:
```diff
@@ -71,6 +71,9 @@
     JC69,
     MatrixMutationModel,
     MutationModel,
+    SMM_MicrosatMutationModel,
```
so `SMM` is probably well enough known on it's own. Many of the others that this generalized system I implemented might not be. E.g. EL2. I doubt it will be _that_ confusing to users however as there will be no name clash with the AA or nucleotide models

andrewkern commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+        u = 0.5
+        v = 0
+        transition_matrix = general_microsat_rate_matrix(s, u, v, p, m, lo, hi)
+        if root_distribution is None:
```
yep that's a better design. I'll make this edit.

petrelharp commented on docs/mutations.md:
```diff
@@ -661,9 +668,33 @@ print(f"Genetic diversity: {theta}.")
+the human genome. Let's simulate a sample of 20 individuals at 13 unlinked loci,
+and then apply the {class}`.SMM_MicrosatMutationModel` model to the sample.
+
+Finally we will summarize variation across our loci
```
fragment sentence here

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+    return max(0, min(1, u - (v * (i - lo))))
+
+
+def general_microsat_rate_matrix(s, u, v, p, m, lo, hi):
```
I've had a look at this, and I guess my vote is to leave the parameters here matching what's in the paper - there's enough of them and it's confusing enough that I like the short names that are easier to compare to the paper.

jeromekelleher replied:
If we're matching the param names in the paper then that's a good justification. Let's keep them this way so.

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
         super().__init__(alleles, root_distribution, transition_matrix)
 
 
+def gamma_microsat(m, i, j, lo, hi):
```
maybe these helper functions should be defined internally to the `general_microsat_rate_matrix` function?

andrewkern replied:
👍

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+    for microsatellite repeats number.
+
+    This is a :class:`.MatrixMutationModel` with alleles ``[lo, .. , hi]``,
+    root distribution is uniform ``(lo, hi)``, and a transition matrix that reflects
```
```suggestion
    a root distribution that is uniform ``(lo, hi)``, and a transition matrix that reflects
```

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+    Stepwise mutation model (Kimura and Ohta, 1978),
+    for microsatellite repeats number.
+
+    This is a :class:`.MatrixMutationModel` with alleles ``[lo, .. , hi]``,
+    root distribution is uniform ``(lo, hi)``, and a transition matrix that reflects
+    single-step mutation.
```
```suggestion
    This is a :class:`.MatrixMutationModel` with alleles ``[lo, .. , hi]``,
    a root distribution that is uniform ``(lo, hi)``, and a transition matrix that allows
    only mutations that change the number of repeats by +/- 1.
```

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+class SMM_MicrosatMutationModel(MatrixMutationModel):
+    """
+    Stepwise mutation model (Kimura and Ohta, 1978),
+    for microsatellite repeats number.
```
```suggestion
    for microsatellite repeat number.
```

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+    root distribution is uniform ``(lo, hi)``, and a transition matrix that reflects
+    single-step mutation.
+
+    Note that the values of ``lo`` and ``hi`` are in units of repeat number,
+    and so can be arbitrarily large with respect to repeat length
+    (e.g. dinucleotide, tri-nucleotide, etc.).
```
```suggestion
    Note that the values of ``lo`` and ``hi`` are in units of repeat number,
    and are unrelated to repeat length (e.g., the repeat itself can be dinucleotide, tri-nucleotide, etcetera).
```
Is this what you meant?

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+
+    :param int lo: lower bound of the microsatellite repeat copy number
+    :param int hi: upper bound of the microsatellite repeat copy number
+    :param list[float] root_distribution: An array of float values of length hi-lo+1
+        which should sum to 1. These values are used to determine the ancestral state of
+        each microsatellite. Defaults to the value of stationary distribution
+        of the model
```
```suggestion
    :param list[float] root_distribution: An array of float values of length hi-lo+1
        which should sum to 1. These values are used to determine the ancestral state of
        each microsatellite. Defaults to the stationary distribution
        of the model.
```

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+    This models evolution of microsatellite repeats number in a manner
+    that allows for multi-step mutations in copy number of microsatellite repeats.
+    This is parameterized by `p` the probability of a single step mutation,
+    and `m` the success probability of the truncated gamma distribution.
+    For this model both `p` and `m` need to be set to < 1.0.
```
```suggestion
    and `m` the success probability of the truncated gamma distribution
    describing the distribution of longer steps.
    For this model both `p` and `m` need to be set to < 1.0.
```

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+    :param list[float] root_distribution: An array of float values of length hi-lo+1
+        which should sum to 1. These values are used to determine the ancestral state of
+        each microsatellite. Defaults to the value of stationary distribution
+        of the model
```
see comment above.

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+
+    This is a :class:`.MatrixMutationModel` with alleles ``[lo, .. , hi]``,
+    root distribution is uniform ``(lo, hi)``, and a transition matrix that reflects
+    single-step mutation.
```
fixup here again

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,285 @@ def __init__(self, state_independent=False):
+
+    Note that the values of ``lo`` and ``hi`` are in units of repeat number,
+    and so can be arbitrarily large with respect to repeat length
+    (e.g. dinucleotide, tri-nucleotide, etc.).
```
ditto

andrewkern commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,37 @@ def test_new_model(self):
+        self.validate_model(model)
+        self.validate_stationary(model)
+
+    def test_EL2_MicrosatMutationModel(self):
```
okay i added a couple tests like this that hit all the bits of the transition matrix that are currently being used. we should add more tests for the `s` param stuff once/if we add those models

andrewkern commented on docs/mutations.md:
````diff
@@ -661,9 +668,33 @@ print(f"Genetic diversity: {theta}.")
+for var in mts.variants():
+    print(var.site.position, var.alleles, var.genotypes, sep="\t")
+```
+
````
done. also added another example on.

andrewkern commented:
okay, I think this is all cleaned up and ready to go. I've added a bunch more tests to verify the transition matrices, cleaned up docs, refactored stuff according to the nice design suggestions (thanks!), and rebased and squashed.

jeromekelleher commented:
LGTM, thanks @andrewkern! Can you update the CHANGELOG please? We're good to merge then.

I'll open an issue to track statistical validation of these sims.

andrewkern commented:
done

jeromekelleher reviewed (commented):
Looks great, thanks!

It's nice to have defaults, but we need to document what they are and have a good justification for setting them to that value. 

Also hard-coding the defuaults into the signatures means that we have no flexibility later in terms of changing the defaults depending on what the other inputs are. The general pattern is to use ``param=None`` in the signature to avoid painting ourselves into corners (unless it's totally clear that there'll never be another other params). Also it avoids having to put the hard coded params into several functions.

However, we can make an issue to track this if you'd rather merge this now?

jeromekelleher commented:
On second thoughts, let's just merge and update. Opened #2074 to track.

petrelharp commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,136 @@ def test_new_model(self):
+                if i != j:
+                    assert (
+                        model.transition_matrix[i, j] == model.transition_matrix[j, i]
+                    )
```
```suggestion
                    )
                if np.abs(i - j) > 1:
                    assert model.transition_matrix[i, j] == 0.0
```

petrelharp commented:
Lots of minor nitpicks (apologies). But: if this is going to be used for simulating CODIS loci I think we should have more exhaustive tests for correctness. Let's see:

- test that alleles are actually between `lo` and `hi` (inclusive)
- test that rates are 0 for non-adjacent ones in SMM
- test that stationary distribution is uniform when we say it is
- write tests for ratios of expressions for EL2 without copy-pasting the helper functions (I've suggested how to do this for TPM
- test that the sum of qij for j > i divided by the sum for j < i is equal to to `(1 - alpha)/alpha`
- test that the sum of `qij` for `i \neq j` divided by the same thing for `i = lo` is equal to `beta`

And, I think that tests everything?

jeromekelleher commented:
OK, removing the AUTOMERGE then since we should capture @petrelharp's comments in this PR.

FYI @andrewkern, I pushed to this branch to update the CHANGELOG date (the date of release was set to today, which might be confusing on the "current" docs build). This means you'll need to fetch from your origin and reset --hard. Or, just force push over the change I made and we can do it again, no biggie.

andrewkern commented on docs/mutations.md:
```diff
@@ -661,9 +668,61 @@ print(f"Genetic diversity: {theta}.")
+
+Consider the [Combined DNA Index System](https://en.wikipedia.org/wiki/Combined_DNA_Index_System) (CODIS), a US National database
+of genotypes maintained by the justice system for use in forensic DNA analysis.
+The CODIS system relies on 13 core microsatellite loci, that are unlinked across
+the human genome. Let's simulate a sample of 20 individuals at 13 unlinked loci,
```
well crap i'm behind the times.

will change this in the simulations as well on my next pass

andrewkern commented on docs/mutations.md:
````diff
@@ -661,9 +668,61 @@ print(f"Genetic diversity: {theta}.")
+```
+
+So here we are seeing a clear relationship between the variance
+in repeat number and mutation rate 🔥🔥🔥
````
🔥

andrewkern commented on msprime/mutations.py:
```diff
@@ -199,6 +199,283 @@ def __init__(self, state_independent=False):
+
+    """
+
+    def __init__(self, p=0.9, m=0.4, u=0.7, v=0.4, lo=2, hi=50, root_distribution=None):
```
these were defaults i was using for testing, loosely inspired by the parameter estimates from Sainudiin et al. (2004).

i think a good way to go for defaults for each of these models would be the exact estimates from that paper

andrewkern commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,136 @@ def test_new_model(self):
+
+        # test for ratio of TPM trans mat vals
+
+        def gamma_microsat(m, i, j, lo, hi):
+            """
+            gamma helper function from Sainudiin et al. (2004)
+            :param float m: prob success
+            :param int i: in-state
+            :param int j: out-state
+            :param int lo: lower bound of microsat repeat length distribution
+            :param int hi: upper bound of microsat repeat length distribution
+            """
+            if lo <= i < j <= hi:
+                num = m * (1 - m) ** (np.abs(i - j) - 1)
+                denom = 1 - (1 - m) ** (hi - i)
+            elif lo <= j < i <= hi:
+                num = m * (1 - m) ** (np.abs(i - j) - 1)
+                denom = 1 - (1 - m) ** (i - lo)
+            else:
+                raise TypeError(
+                    "microsat model error. i and j must be between lo and hi"
+                )
+
+            return num / denom
```
ah yeah this is cleaner

andrewkern commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,136 @@ def test_new_model(self):
+                    model.transition_matrix[ii, ii + 1]
+                    / model.transition_matrix[ii, ii - 1],
+                )
+                # this test might seem a little loosey goosey
+                # but the scaling is create numerical noise i think
+                assert np.abs(exp - obs) < 1e-2
```
i will recheck this

andrewkern commented:
> Lots of minor nitpicks (apologies). But: if this is going to be used for simulating CODIS loci I think we should have more exhaustive tests for correctness. Let's see:
> 
> * test that alleles are actually between `lo` and `hi` (inclusive)
> * test that rates are 0 for non-adjacent ones in SMM
> * test that stationary distribution is uniform when we say it is
> * write tests for ratios of expressions for EL2 without copy-pasting the helper functions (I've suggested how to do this for TPM
> * test that the sum of qij for j > i divided by the sum for j < i is equal to to `(1 - alpha)/alpha`
> * test that the sum of `qij` for `i \neq j` divided by the same thing for `i = lo` is equal to `beta`
> 
> And, I think that tests everything?

Okay I think I've not got all these tests in, EXCEPT for the last one which I'm confused on-- I don't think that's true for all models where s = 0? might need coffee

petrelharp commented on tests/test_mutations.py:
```diff
@@ -412,6 +412,39 @@ def test_EL2(self):
+            assert (d[0] / d[j]) - (
+                betas[0] / betas[j] * alphas[0] / (1 - alphas[j])
+            ) < 1e-6
+        # 2nd parameterization, slight variation
```
Glad you're doing multiple parameterizations - perhaps make sure to set `p` not equal to zero in one of them? And it'd be nicer to do this with `@pytest.mark.parametrize`?

andrewkern replied:
will do

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,301 @@ def __init__(self, state_independent=False):
+        hi=50,
+        root_distribution=None,
+    ):
+        if np.isinf(lo):
+            raise ValueError("lo is infinite")
```
Should probably check that `lo >= 0` and that `lo < hi`?

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,301 @@ def __init__(self, state_independent=False):
+        :param int lo: lower bound
+        :param int hi: upper bound
+        """
+        return 1 + ((i - lo) * s)
```
Bad things happen if `s < -1/(hi - lo)`; we should probably check for and error in this case? (gee, if `s = -1/(hi-lo)` then `hi` is an absorbing state, and the eigenvalue computation probably fails?)

petrelharp replied:
Ah in Sainudiin they say that `s > - 1/(hi - lo + 1)`.

andrewkern replied:
okay i'll add this as a specific test in `MicrosatMutationModel`

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,301 @@ def __init__(self, state_independent=False):
+
+    :param float s: strength of length dependence on mutation rate
+        (i.e. s=0 uniform rate)
+    :param float u: constant bias parameter; must be between 0 and 1
```
We should probably raise an error if u is not between 0 and 1?

petrelharp commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,119 @@ def test_new_model(self):
+                    model.transition_matrix[ii, ii + 1]
+                    / model.transition_matrix[ii, ii - 1]
+                )
+                if np.abs(exp - obs) > 1e-6:
+                    raise ValueError(f"failed test on row: {i} exp:{exp} obs:{obs}")
```
Whoops - left over from debugging?
```suggestion
                assert np.abs(exp - obs) < 1e-6
```

andrewkern replied:
yup

petrelharp commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,119 @@ def test_new_model(self):
+            (0.43, 0.68, 0.037, 1.0, 0.88),
+        ],
+    )
+    def test_MicrosatMutationModel(self, m, u, v, p, s):
```
love it - looks great.

petrelharp commented on tests/test_mutations.py:
```diff
@@ -339,6 +339,119 @@ def test_new_model(self):
+                        model.transition_matrix[i, j] == model.transition_matrix[j, i]
+                    )
+                if np.abs(i - j) > 1:
+                    assert model.transition_matrix[i, j] == 0.0
```
```suggestion
                    assert model.transition_matrix[i, j] == 0.0
                elif np.abs(i - j) == 1:
                    assert model.transition_matrix[i, j] == model.transition_matrix[0, 1]
```

petrelharp reviewed (approved):
LGTM!! - just a couple of minor things (aren't checking for validity of a few parameters).

andrewkern commented:
okay this should be all set i think @petrelharp @jeromekelleher? 

note I haven't edited the CHANGELOG further. not sure if i need to?

andrewkern commented:
well darn now i need to hit these new features tests with tests... lol. brb

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,283 @@ def __init__(self, state_independent=False):
+    :math:`0 < u < 1` and a linear bias parameter `v`, where
+    :math:`-\\infty < v < \\infty`.
+
+    This is a :class:`.MicrosatMutationModel` with alleles ``[lo, .. , hi]``
```
What about this suggestion?

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,308 @@ def __init__(self, state_independent=False):
+            raise ValueError("lo must be >= 0")
+        if s <= -1 / (hi - lo + 1):
+            raise ValueError("s must be > -1/(hi-lo+1)")
+        if u < 0 or u > 1:
+            raise ValueError("u must be between 0 and 1 inclusive")
```
I think this should be *non*-inclusive. Otherwise, some wierd things can happen, e.g. the chain is not ergodic:
```
>>> mm = msprime.MicrosatMutationModel(u=0, lo=3, hi=6, v=-10)
>>> mm.transition_matrix
array([[1., 0., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 1.]])
>>> mm.root_distribution
array([1., 0., 0., 0.])

```

andrewkern replied:
we can change this but i think it conflicts with the original paper
<img width="483" alt="image" src="https://user-images.githubusercontent.com/7674197/175462652-dd9262bc-2881-440d-a123-215c9435a629.png">

petrelharp replied:
Well, you get a valid Markov chain for the endpoints, but where one or both of the endpoints are absorbing, and so you'd get a situation where the root distribution draws with probability 1 from an absorbing state, and so you'd never get any non-silent mutations. That seems like not what anyone would actually want? And, the derived models have 0 < u < 1? There's no harm in leaving it 0 <= u <= 1, I guess, but I'd vote to restrict it.

petrelharp commented:
Terribly sorry, but I have one more change (I think it should be 0 < u < 1 not <=), and there's two pending suggestions?

andrewkern commented:
pending suggestions should be all set in my commits-- i didn't use the GH interface to incorporate them, but let me know if i missed something

petrelharp commented on msprime/mutations.py:
```diff
@@ -199,6 +199,308 @@ def __init__(self, state_independent=False):
+    :param int lo: lower bound of the microsatellite repeat copy number
+    :param int hi: upper bound of the microsatellite repeat copy number
+    :param list[float] root_distribution: An array of float values of length hi-lo+1
+        which should sum to 1. These values are used to determine the ancestral state of
+        each microsatellite. Defaults to the value of stationary distribution
+        of the model
```
This is what I meant here:
```suggestion
        which should sum to 1, and give the probabilities used to choose
        the ancestral state of each microsatellite.
        Defaults to the value of stationary distribution
        of the model.
```

petrelharp commented:
Here's the two missing ones. I think they're important because they say what precisely what the mutation rates are.
![Screenshot from 2022-06-23 21-58-11](https://user-images.githubusercontent.com/1046249/175465469-283e2310-9f73-490c-9b79-ae7a16bdca77.png)
![Screenshot from 2022-06-23 21-57-47](https://user-images.githubusercontent.com/1046249/175465472-72315c5a-236c-4701-b90e-e9d979d80fb8.png)

petrelharp commented:
Oh! And another important thing - where did the defaults come from?

andrewkern commented:
okay got the suggestions. will post a separate commit for limiting u

andrewkern commented:
okay got a commit in that will hopefully close #2074?

petrelharp commented on msprime/mutations.py:
```diff
@@ -471,8 +471,8 @@ class EL2(MicrosatMutationModel):
 
     This is a :class:`.MicrosatMutationModel` with alleles ``[lo, .. , hi]``
-    For a precise definition of the mutation rates, see :class:`.MicrosatMutationModel`
-    with ``s=0``.
```
(same as above)

andrewkern commented on msprime/mutations.py:
```diff
@@ -422,8 +422,8 @@ class TPM(MicrosatMutationModel):
 
     This is a :class:`.MicrosatMutationModel` with alleles ``[lo, .. , hi]``.
-    For a precise definition of the mutation rates, see :class:`.MicrosatMutationModel`
-    with ``s=0``, ``u=0.5``, and ``v=0``.
```
Sorry. I thought it was more precise to say `Default parameterization as estimated in Sainudiin et al. (2004).` but I'll keep in this bit above as well as my addition
